enhancements/substitutes-for.md

```diff
@@ -0,0 +1,381 @@
+If a new bundle is added to the middle of the graph, replaces/skips references must be updated to include
+the new node.
+
+There is a need to rebuild old versions of bundles to pick up CVE fixes.  Doing so requires that the
+upgrade graph is preserved such that users who update to the CVE-patched bundle can still continue to
+upgrade to newer bundles, just as if they were still on the un-patched bundle.
+
+For example, suppose the original upgrade graph, as specified via "replaces" references, was:
+
+1.0.0->1.0.1
+
+If the patched version is 1.0.1-patched (due to semver semantics this is ordered before 1.0.1), then ideally we want
+the new graph to be:
+
+[1.0.0 or 1.0.1-patched]->1.0.1
+
+However achieving such a graph would require updating the 1.0.1 bundle to add a "replaces 1.0.1-patched" reference.
+In turn, 1.0.1 can't be updated without changing its version, which means we end up needing something like:
+
+[1.0.0 or 1.0.1-patched or 1.0.1]->1.0.2
+
+(Where the only difference between 1.0.1 and 1.0.2 is the replaces references).
+
+
+To avoid having to update and publish new versions of subsequent bundles in the upgrade graph, we need a way to
+effectively stitch 1.0.1-patched into the graph without changing 1.0.1 (or any other existing bundles that need
+to be able to upgrade from 1.0.1-patched).
+
+This enhancement proposes we do so by allowing a bundle to specify which other bundles it "substitutes for".  This
+will effectively act as a search/replace in the existing bundles for any references to the bundles this bundle 
+substitutes for.  
+
+Continuing the example above, if 1.0.1-patch is annotated as substituting for 1.0.0, then the effective new upgrade 
+graph would be:
+
+1.0.0->1.0.0-patched->1.0.1
+
+
+The implications of this for opm add + the catalog db are that the substitutes-for metadata must be stored in the DB associated w/ each bundle.
+
+The expected flow is:
+
+1) opm add inserts the new bundle following all existing logic, but also adding the substitutes-for metadata
+2) Upon a list-bundles request, the operator-registry should return a substitution-corrected list of bundles based on substitutes replaces/skips references.
+3) In addition, the registry export command will need to perform substitutions on the CSVs as it exports them, to ensure their replaces/skips references are correct.
+
+
+Complications:
+
+1) the CSV itself contains other version references(beyond skips/replaces), but it should be ok not to perform updates on them because those references are not used by anything making upgrade decisions
+
+2) skipRange logic could theoretically get invalidated (1.2 substitutes for 1.1, but some bundle's skipRange is "skipRange: <1.2").  Since this feature is primarily being added for CVE rebuilds which we expect to only bump the patch release, this should simply be documented as a known limitation/caveat around using substitutes-for (that it does not allow substitution consideration during skipRange upgrades)
```

Comment:
for "pure" skipRange users, the CVE rebuild need not reference the previous build in any way and the next build need not reference the CVE rebuild. So we need to find a version bump convention that:
1. preserves the skipRange of the CVE
2. maintains updates from the previous versions to the patched version

so If I have 1.0.0 with skipRange < 1.0.0 and create 1.0.0-patched and keep the skipRange as <1.0.0 then now the skipRange includes the patched version...

Reply:
nevermind 1.0.1-patched with skipRange preserved works

Reply:
I'm confused as to what the caveat is here sorry

Reply:
the caveat is around cases where people use substitutes-for across minor versions, not patch versions, such that they might end up with "X substitutes for Y" but while Y falls within some existing skiprange, X does not.  So even though X substitutes for Y, users will not be able to upgrade from X to HEAD, because HEAD's skipRange does not include X.

There is nothing we can really do about it.  The only solution to that is to update HEAD's skiprange (or update HEAD's replaces/skips), which of course requires a new version of HEAD, which we've been trying to avoid.

Arguably it could happen within a patch version too, if you have tightly constrained skipRanges.

Reply:
skipranges are also stored separately in the database and can be updated (much like skips/replaces).

I would think that the way this should work for adding a X' that substitutes-for X:

1. find all csvs for which X falls in the skiprange
2. update the skiprange to include X'
3. (because this is semver range, this can just be explicit, i.e. tack on a `, X'` to the end of the skiprange string

Reply:
> since the skipRange of channel heads are the only ones that impact upgrades


that's no longer true in 4.6+, the resolver will consider skipranges throughout